File: leetcode/0102.binary-tree-level-order-traversal/javascript/001.bfs.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
  const levels = [];

  if(!root) return levels;

  const queue = [root];

  while(queue.length) {
    const level = [];
 
    for(let i = queue.length; i > 0; i--) {
      const node = queue.pop();

      if(node !== null) {
        level.push(node.val);
        queue.unshift(node.left);
        queue.unshift(node.right);  
      }
    }

    if(level.length) levels.push(level);
  }

  return levels;
};

File: leetcode/0104.maximum-depth-of-binary-tree/go/001.bfs.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
    depth := 0

    if root == nil {
        return depth
    }

    queue := []*TreeNode{root}

    for len(queue) > 0 {
        levelQueueSize := len(queue)

        for i := 0; i < levelQueueSize; i += 1 {
            node := queue[0]
            queue = queue[1:]

            if node.Left != nil {
                queue = append(queue, node.Left)
            }
            
            if node.Right != nil {
                queue = append(queue, node.Right)
            }
        }

        depth += 1
    }

    return depth
}
